compiler/docgen.nim

```diff
@@ -1347,7 +1347,7 @@ proc commandRstAux(cache: IdentCache, conf: ConfigRef;
   var d = newDocumentor(filen, cache, conf, outExt)
 
   d.isPureRst = true
-  var rst = parseRst(readFile(filen.string), filen.string, 0, 1, d.hasToc,
+  var rst = parseRst(readFile(filen.string), filen.string, 1, 1, d.hasToc,
```

Comment:
not sure those line changes are correct, because other compiler code might already do the offset:
```nim
ok1

.. asdfasdf::

ok2

```
before PR:
t11958.rst(3, 2) Error: invalid directive: 'asdfasdf'
after PR:
t11958.rst(4, 2) Error: invalid directive: 'asdfasdf'
=> incorrect line number

in future work, maybe we should add a `type LineNumber = distinct int` (1-based) or similar to avoid guesswork everywhere

Reply:
I checked the case of directives thoroughly: it happens because `getDirective` https://github.com/nim-lang/Nim/blob/6c1931747aece5f95f6d67876a558d05940d4547/lib/packages/docutils/rst.nim#L1282 
advances `p.idx` to the next line.

I fixed it. Also I remembered clearly that indeed all elements errors/warnings had the line shifted by `-1`, except directives, that had it right.

Your 2 other ideas are good as usual... Especially it's worth checking line numbers in the test. Currently we have a more basic problem that I haven't learned how to handle `rstMessage` invocation in test, it just throws `AssertionDeffect` by some reason, which is ridiculous :-(